packages/frontend/workspace/src/providers/index.ts

```diff
@@ -1,160 +1,114 @@
+import { createIndexedDBStorage } from './storage/indexeddb';
+import { createSQLiteStorage } from './storage/sqlite';
+import { SyncEngine } from './sync';
 
-const createIndexedDBBackgroundProvider: DocProviderCreator = (
-  id,
-  blockSuiteWorkspace
-): LocalIndexedDBBackgroundProvider => {
-  const indexeddbProvider = create(blockSuiteWorkspace);
-
-  let connected = false;
-  return {
-    flavour: 'local-indexeddb-background',
-    datasource: indexeddbProvider.datasource,
-    passive: true,
-    get status() {
-      return indexeddbProvider.status;
-    },
-    subscribeStatusChange: indexeddbProvider.subscribeStatusChange,
-    get connected() {
-      return connected;
-    },
-    cleanup: () => {
-      indexeddbProvider.cleanup().catch(console.error);
-    },
-    connect: () => {
-      logger.info('connect indexeddb provider', id);
-      indexeddbProvider.connect();
-    },
-    disconnect: () => {
-      assertExists(indexeddbProvider);
-      logger.info('disconnect indexeddb provider', id);
-      indexeddbProvider.disconnect();
-      connected = false;
-    },
-  };
-};
-
-const indexedDBDownloadOrigin = 'indexeddb-download-provider';
-
-const createIndexedDBDownloadProvider: DocProviderCreator = (
-  id,
-  doc
-): LocalIndexedDBDownloadProvider => {
-  const datasource = createIndexedDBDatasource({});
-  let _resolve: () => void;
-  let _reject: (error: unknown) => void;
-  const promise = new Promise<void>((resolve, reject) => {
-    _resolve = resolve;
-    _reject = reject;
-  });
-
-  return {
-    flavour: 'local-indexeddb',
-    active: true,
-    get whenReady() {
-      return promise;
-    },
-    cleanup: () => {
-      // todo: cleanup data
-    },
-    sync: () => {
-      logger.info('sync indexeddb provider', id);
-      datasource
-        .queryDocState(doc.guid, {
-          stateVector: encodeStateVector(doc),
-        })
-        .then(docState => {
-          if (docState) {
-            Y.applyUpdate(doc, docState.missing, indexedDBDownloadOrigin);
-          }
-          _resolve();
-        })
-        .catch(_reject);
-    },
-  };
-};
-
-export {
-  createAffineSocketIOProvider,
-  createBroadcastChannelProvider,
-  createIndexedDBBackgroundProvider,
-  createIndexedDBDownloadProvider,
-  createSQLiteDBDownloadProvider,
-  createSQLiteProvider,
-  downloadBinaryFromCloud,
-};
+export * from './sync';
 
 export const createLocalProviders = (): DocProviderCreator[] => {
-  const providers = [
-    createIndexedDBBackgroundProvider,
-    createIndexedDBDownloadProvider,
-  ] as DocProviderCreator[];
-
-  if (runtimeConfig.enableBroadcastChannelProvider) {
-    providers.push(createBroadcastChannelProvider);
-  }
-
-  if (environment.isDesktop && runtimeConfig.enableSQLiteProvider) {
-    providers.push(createSQLiteProvider, createSQLiteDBDownloadProvider);
-  }
-
-  return providers;
+  return [
+    (_, doc, { awareness }) => {
+      const engine = new SyncEngine(
```

Comment:
I prefer moving `Doc` as first parameter of constructor of `SyncEngine` because the doc is the first thing that makes sync happen